### DISCUSSION TITLE
Open gopro for raspberry pi ?

### DISCUSSION THREAD

**Discussion by roberto1963** [gopro/OpenGoPro#27] [Q&A] [answered]
Hello , 
I would like to control a gopro hero9 with bluetooth and a raspberry Pi.
I have install open gopro on a Raspberry Pi3+ .
I have follow instructions of https://gopro.github.io/OpenGoPro/python_sdk/  (but with pip3 instead of pip) . 
But I am unable to use it.
Raspberry Pi python3 version is only Python 3.7.3  (not 3.8) .   Is it a problem ?
Has anyone tried open gopro with a raspberry pi ? 
Thanks.
Roberto.

**Comment by tcamise-gpsw** [comment]
Hello. Yes, you do need the correct python version: 3.8.x. 

Other than that, the python SDK has limited testing on Linux. I have seen at least one bug in the bleak driver where it can not pair. In order to get around this, I had to manually connect and pair first using BlueZ from the command. After that, the python SDK does work as expected.

I started a [discussion ](https://github.com/hbldh/bleak/discussions/503) about this on the bleak repo but haven't gotten around to filing an issue / debugging it more yet.

I haven't tested this in months so perhaps it is fixed now. In any case, I encourage you to give it a try and report your findings 👍

**Comment by KonradIT** [comment]
Here's some more info on GoPro BLE on the Pi: https://github.com/KonradIT/goprowifihack/blob/master/Bluetooth/Platforms/RaspberryPi.md

Those instructions are as bulletproof it gets for pairing Linux/RPI with GoPro's BLE. I recommend everyone follows before using the demos here.

I'm able to control the camera using the CLI tool `gatttool`, my own gopro-ble-py with no issues. gopro-ble-py also uses bleak for the BLE backend.

![image](https://user-images.githubusercontent.com/5891284/127736317-c52df185-65a5-4748-b182-f1fdd9c6eeaf.png)

This is on a Raspberry Pi 4, with BlueZ version 5.50. To pair the camera I followed the steps on the guide I linked earlier, it seems no other way works.

Note that I used the gopro-ble-py's Pipfile, which forces Python 3.8 and also installs the bleak library. Maybe OpenGoPro could have a pipfile as well, as it would make installing the deps easier.

Unable to try OpenGoPro due to the program connecting to GoPro's WiFi thus kicking me out of the shell.

**Reply by tcamise-gpsw** [reply]
@KonradIT Great info here! Are you saying that you are able to use bleak on the RPi (via your gopro-ble-py) without first pairing via the CLI? I have not been able to accomplish this from Ubuntu so that would be great news.

Also, regarding the OpenGoPro stuff:
- specifying version: This is a good idea and I think I should be able to achieve this in the setup.py. Pipfile is specific to pipenv, correct?
- WiFi: I have a PR open right now to make the WiFI connection optional. I'm still adding some more stuff but feel free to review if you'd like.

**Reply by KonradIT** [reply]
Hi, I had to first pair via CLI using bluetoothctl (and use gatttool also to check everything is ok), then I could use bleak.

Pipfile is specific to pipenv

I will check the branch as well and try on the Pi.

**Reply by tcamise-gpsw** [reply]
Ok yeah that's the same behavior I see. I'm going to try to dig into this in bleak this week. It certainly seems like it should be possible without using the CLI.

**Reply by tcamise-gpsw** [reply]
Great news! It looks like there is already an open PR on bleak to add this functionality: https://github.com/hbldh/bleak/pull/523

**Reply by SophieLangdon27** [reply]
Hi, this is a bit late but I've managed to connect my GoPro through bluetoothctl. However, I'm not sure how you could run a python script and use bleak from here as I'm now stuck in the Bluetooth interface and can't call python commands. I was wondering how you managed this?

**Comment by tcamise-gpsw** [comment, marked as answer]
Marking this an answered. Per the discussion above, it is possible to use the Open GoPro Python SDK with the RPi. However, due to a current Bleak limitation, you must first pair from the command line using BlueZ. There is future work to improve this which can be tracked here: https://github.com/gopro/OpenGoPro/issues/29
